### PR TITLE
Fix some threading issues.

### PR DESCRIPTION
**Pull request by AlCutter**

### CODE REVIEW

File: cpp/server/ct-server.cc [hunk 1]
```diff
@@ -288,14 +292,46 @@ void WatchdogTimeout(int sig) {
...
+}  // namespace
+
+
 int main(int argc, char* argv[]) {
   google::ParseCommandLineFlags(&argc, &argv, true);
   google::InitGoogleLogging(argv[0]);
   google::InstallFailureSignalHandler();
+
+  // Set-up OpenSSL for multithreaded use:
+  {
```
Comment (pphaneuf) [inline cpp/server/ct-server.cc]: Why the braces?

File: cpp/server/ct-server.cc [hunk 2]
```diff
@@ -288,14 +292,46 @@ void WatchdogTimeout(int sig) {
...
+    CRYPTO_THREADID_set_callback(threadid_function);
+    CRYPTO_set_locking_callback(locking_function);
+  }
+
   OpenSSL_add_all_algorithms();
   ERR_load_crypto_strings();
   cert_trans::LoadCtExtensions();
 
+
```
Comment (pphaneuf) [inline cpp/server/ct-server.cc]: Spurious.

File: cpp/server/ct-server.cc [hunk 3]
```diff
@@ -117,11 +119,13 @@ using std::chrono::seconds;
 using std::chrono::steady_clock;
 using std::function;
 using std::make_shared;
+using std::mutex;
 using std::placeholders::_1;
 using std::shared_ptr;
 using std::string;
 using std::thread;
 using std::unique_ptr;
+using std::vector;
```
Comment (pphaneuf) [inline cpp/server/ct-server.cc]: Looks like a possible leftover? Tried to use a static vector in ```locking_function```, perhaps?

File: cpp/log/frontend.cc [hunk 4]
```diff
@@ -24,6 +26,7 @@ Frontend::~Frontend() {
 }
 
 void Frontend::GetStats(Frontend::FrontendStats* stats) const {
+  lock_guard<mutex> lock(stats_mutex_);
```
Comment (pphaneuf) [inline cpp/log/frontend.cc]: You're a generous man, I would have been tempted to remove the (unused) stats instead... :wink:

**Comment by pphaneuf** [comment]
That's one weird Travis failure!

File: cpp/server/ct-server.cc (hunk 1 above)
Comment (AlCutter) [inline cpp/server/ct-server.cc]: There was a bunch more lines in there originally, including the vector of locks which I was going to bind in, before realising this is old school and had to change it all :)
Dinged 'em.

File: cpp/server/ct-server.cc (hunk 2 above)
Comment (AlCutter) [inline cpp/server/ct-server.cc]: Despuriated.

File: cpp/server/ct-server.cc (hunk 3 above)
Comment (AlCutter) [inline cpp/server/ct-server.cc]: gone.

File: cpp/log/frontend.cc (hunk 4 above)
Comment (AlCutter) [inline cpp/log/frontend.cc]: it was quicker to do this, and anyway, maybe I'll convert these to monitoring stats one day :)

**Comment by pphaneuf** [comment]
LGTM

**Comment by pphaneuf** [comment]
Looks like an actual problem, I'm guessing something related to your changes in ```cpp/Makefile```?

**Comment by pphaneuf** [comment]
Merged as cefe6667f19fda5b53d3f2c0553df062fb6f97ed.

**Pull request closed without merging**